### PR TITLE
Modified the model_serialization to have correct inputs and outputs

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Before : 
Screen Shot 2021-02-26 at 16 03 43
And after :
Screen Shot 2021-02-26 at 16 03 52

This should not be a breaking change since the model uses the Barracuda implicit input and output names. This will be useful when Barracuda will allow us to use named tensors for LSTM.

EDIT:
Some context. Currently, Barracuda creates new names for inputs and outputs of the LSTM (see here https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/Runtime/Inference/TensorApplier.cs#L79) and we use these at inference. This means that neither RecurrentOutput nor RecurrentOutputH/C are used at all. When Barracuda switches to use the names of the ONNX model, we need to make sure that there are appropriately named inputs and outputs to the LSTM.

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

Slack message : https://unity.slack.com/archives/C8FECS6L9/p1614370112116900
JIRA : MLA-1807

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/torch/model_serialization.py [hunk 1]
```diff
@@ -40,6 +40,36 @@ def is_exporting():
         return exporting_to_onnx._local_data._is_exporting
 
 
+class TensorNames:
+    BatchSizePlaceholder = "batch_size"
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/torch/model_serialization.py:44]: Please follow python naming conventions for these. I understand that you probably wanted to copy-paste the C# code here, but it makes the resulting python code stick out.

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/model_serialization.py:44]: Ok, will do. 
Just to make sure, are we okay with this approach of having a class with static strings or is there a more Pythonic way to do this?

Comment (chriselion) [reply ml-agents/mlagents/trainers/torch/model_serialization.py:44]: You could do it as a enum too (python enums aren't constrained to integer values) - see `BufferKey` in buffer.py, but there's not much advantage to doing that here (and you'd have to write a bit more code to cast to strings).

You could also add methods to the class for combining the prefix + index, like `TensorNames.get_visual_observation_name(i)`

File: ml-agents/mlagents/trainers/torch/model_serialization.py [hunk 2]
```diff
@@ -40,6 +40,36 @@ def is_exporting():
...
+    VisualObservationPlaceholderPrefix = "visual_observation_"
+    ObservationPlaceholderPrefix = "obs_"
+    PreviousActionPlaceholder = "prev_action"
+    ActionMaskPlaceholder = "action_masks"
+    RandomNormalEpsilonPlaceholder = "epsilon"
+
+    ValueEstimateOutput = "value_estimate"
+    RecurrentOutput = "recurrent_out"
+    recurrentOutputH = "recurrent_out_h"
+    recurrentOutputC = "recurrent_out_c"
```
Comment (dongruoping) [inline ml-agents/mlagents/trainers/torch/model_serialization.py:59]: Are these used anywhere?

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: No, but I think there is a benefit in having symmetry with the C# code that does the same thing. I can remove them if someone disagrees strongly with it.

File: ml-agents/mlagents/trainers/torch/model_serialization.py (hunk 1 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: I wanted to have as much as possible the same code between C# and Python for the TensorNames, but I think adding helper methods would help.

File: ml-agents/mlagents/trainers/torch/model_serialization.py (hunk 2 above)
Comment (dongruoping) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: Correct me if I'm wrong but seems like in C# `recurrentOutputH` and `recurrentOutputC` are not used either.

File: ml-agents/mlagents/trainers/torch/model_serialization.py (hunk 1 above)
Comment (chriselion) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: > I think adding helper methods would help.

It's right there in the name :)

Fee free to add to C# too.

File: ml-agents/mlagents/trainers/torch/model_serialization.py (hunk 2 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: They are. They are actually the only ones used. recurrentOutput is not used.

Comment (dongruoping) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: TensorGenerator and TensorApplier are both using TensorNames.RecurrentOutput but I don't see where they're using recurrentOutputH? same for recurrent input

File: ml-agents/mlagents/trainers/torch/model_serialization.py (hunk 1 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: done

File: ml-agents/mlagents/trainers/torch/model_serialization.py (hunk 2 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: To clarify, the names used are `recurrent_out_h` and `recurrent_out_c` but these are not grabbed from the TensorNames but directly from Barracuda. But you are right, it does not make sense to keep them explicitly here.

Comment (dongruoping) [reply ml-agents/mlagents/trainers/torch/model_serialization.py]: Thanks for clarifying, that makes sense to me now.

**Review by dongruoping — APPROVED** [review]

**Pull request merged**